### ISSUE TITLE
Publishing HTML on the blockchain

### ISSUE THREAD

**Issue by pfraze** [ssbc/phoenix#400]
This is a mechanism for pages and applications. It follows closely with the existing Web model: you browse HTML pages via links, and the current page occupies the entire window. The main difference is, rather than hosting the pages, you publish them on the blockchain.

Three areas to address: UI, namespacing, and security policies.

**UI**

 - I see no reason not to emulate browsers, which Patchwork already does with a navigation chrome (home, back, forward, reload, address-bar).
 - Pages will be loaded in an Electron `<webview>`, which executes in a separate JS context and process from the containing page.
 - In the future, we can explore ways of breaking out of this model, to create more "desktop app like" experiences.

**Namespacing**

 - This is asking, how do users reference pages that have been published - eg, to share a link - and how do developers reference assets from the HTML?
 - The most direct and secure approach is to reference by the blobs' hash IDs. This should always be available, for situations where the user/dev wants the integrity check.
 - Reasons to introduce a name->hash mapping:
   - Convenience. It's always nicer to have human-readable names. Developers will have to upload all assets before referencing them, which doesnt sound like a feasible flow in some cases.
   - Mutability. Name mappings can be updated to point to new hashes.
 - Some approaches to namespacing:
   1. Give each user a single namespace, which they can manually manage. [This could be presented with a "cloud folder" abstraction.](https://cloud.githubusercontent.com/assets/1270099/8756533/8a1a4542-2c98-11e5-900d-087eba9a8bf9.png)
   2. Let the user publish individual namespaces. This would be similar to posting archives: the namespaces would be self-contained, and detached from any larger namespace. Nice thing about this is, the user doesn't have to garden their "cloud folder" in quite the same way. It also provides a boundary around a set of assets, which could be useful for security reasons. A reference, in this scenario, would either be relative (for assets referencing each other within the archive) or absolute, in which case it's prefixed by the archive's hash. This might also be achieved with existing archive formats: zip, tar.gz, rar. This approach doesn't create mutable mappings.

**Security policies**

 - Similar to the web's hosting domains, there are some natural security domains in SSB: feeds/users, and, if we use approach 2 to namespacing, individual archives. We can also overlay other kinds of security domains, eg sets of users, or capability-token-defined groups. It's up to us how to assign trust to these domains.
 - We should setup some goals. Here are my recommendations:
   1. Unless permitted by user, no data or metadata should leave the page's sandbox.
   2. Unless permitted by user, writes to disk or to the feed, and access to any special resource, should not be allowed.
   3. Irrevocable actions (such as writing to the feed) should always be subject to review before occurring, unless the user sets an "always publish without review" policy.
   4. New versions of a page should always be reviewed by the user before resuming permission policies, if just to make them aware of it.
   5. User comments, audits and ratings should be discoverable within the permission UIs, to help the user make security decisions.
   6. We should allow the most freedom within these constraints.
 - The main tool for controlling the sandbox is the Content Security Policy, which we must set with a response header. Because electron doesn't current allow setting response headers through the protocol-intercept API, we'll need to create an HTTP server for blobs to do this.
 - I would suggest, for v1, the blob server follow policies which allows an HTML-blob, when rendered, to read any other plaintext blob in the local cache, but do nothing else. The policies would be:
   - Only respond to requests from the localhost.
   - Only serve blobs which were published in plaintext (unencrypted).
   - Only respond to GET requests.
   - Do not allow cross-origin requests.
   - Set CSPs to allow free reads within the localhost's domain only.
     - Do not allow plugins, such as Java, to be executed.
   - Do not attempt to locate a blob which the page requested, but could not find.
 - I would also suggest, for v1, injecting an sbot RPC connection into the page (`window.ssb`) which allows only read methods of plaintext data. This will enable the pages to render information and media which the user has locally available.
   - Suggested methods: get, getPublicKey, whoami, relatedMessages, createFeedStream, createUserStream, createLogStream, messagesByType, links
   - This connection should be rate-limited to avoid DoS attacks.
   - In the future, we can add methods which allow publishing on the feed, and possibly access to other resources.

**Comment by dominictarr** [comment]
Hmm, okay this seems like it's two or three things to be honest.
mutable namespaces, reliable sandboxing, and editor interface?

That is fair enough, but we have explored this area before but nothing has stuck because we havn't managed to pull of sandboxing in a satisfying way...

I know we could do the naming part, i think sandboxing is the biggest unknown here, so therefor that aspect should be explored first. Also, we don't actually require the namespaces for this, at least not to demonstrate the sandboxing idea, because we could 1) just use the thing that https://github.com/substack/htmlbin uses, (which bundles a static page into a single html file).

**Comment by dominictarr** [comment]
hmm, also if the entire app was contained in a single bundle then resharing it would be possible too.

**Comment by pfraze** [comment]
We don't need a complex editor interface, I think some way just to open pages that aren't published yet.

I agree about the namespaces. We can get some ways without them, and Im also leaning more towards the second approach, of bundles/archives. Resharing with one hash is possible, fits nicely with a package-manager, can compress the archive also.

**Comment by dominictarr** [comment]
```
<domanic> pfraze, it seems to me that electron sandbox could/should be it's own module
<domanic> it's not really tied to pw at all, except that the api is passed in somewhere
<pfraze> domanic: not sure there's anything to split out. It'll be a combination of using <webview> and the http server, I believe
<domanic> oh but you have a *long* list of settings
<pfraze> yeah, in my issue?
<domanic> yeah...
<domanic> oh how do you init a webview?
<pfraze> https://github.com/atom/electron/blob/master/docs/api/web-view-tag.md
<pfraze> those settings will mostly come down to construction of the server and the CSP header
<domanic> oh so you'd just serve the webview with the right CSP
<domanic> and you are good
<pfraze> yeah
<pfraze> then the RPC will occur over the postMessage interface with the parent frame
<domanic> so the app just needs to include a client lib
<domanic> or inject that or something
* tictacjoe has quit (Remote host closed the connection)
<pfraze> we can inject that, if we like, there are APIs to do that
```
I am happy with this plan.

**Comment by pfraze** [comment]
Cool, I'll start without namespaces and get the pages loading from disk (for dev) and from the feed. Then we can work out the namespace.

**Comment by pfraze** [comment]
First step seems to be working well. I set up two servers: 7777 serves blobs, 7778 serves files on the HD. I could've used `file:` instead of the second server, but I wanted to control the CSP header there too, so that developers know what policies/behaviors to expect.

The test page goes from click to render in, what looks like, half a second, which is good.

Im getting some slight weirdness with the devtools in the webview. It generally works, but it doesn't embed in the webview (it pops out) and it doesnt seem to show any of the requests in the network panel. Will see how that goes.

**Comment by pfraze** [comment]
![screen shot 2015-07-19 at 1 05 56 pm](https://cloud.githubusercontent.com/assets/1270099/8767104/103f449a-2e17-11e5-9cfa-978a7573c3c2.png)

This is progressing well. Can now open files and published blobs in a webview. CSPs are working, RPC is working, devtools are working.

You can include published assets by hash, though I need to fix some edgecase bugs:

```html
<!-- bootstrap 3.3.5 minified -->
<link rel="stylesheet" href="MfvZlkHCEqatNoGiOXveE8FIwMzZg4W85qfrfIFBfYc=.sha256">
```

Some things that would be neat:

 - A button to show all messages which link to the current blob.
 - A button to share the blob on your feed, or via secret message. (This would need to be aware of the hashes of all dependent assets, in order to share it correctly.)

**Comment by dominictarr** [comment]
not having to separate dependant hashes is one thing that would make bundling easier.
another approach would be to parse once and publish all the required objects in the app message,
like metadata.

**Comment by pfraze** [comment]
Yeah, would not be hard (at all) to parse out hash references from the html

**Comment by dominictarr** [comment]
i'm still in favor of bundling it though - it's simpler, and whether or not it will work becomes a single factor.
(you have one blob, or you don't) not that i am against it in the long run, there may certainly be cases where it is better, but i don't think we need it to validate the idea of html apps

**Comment by pfraze** [comment]
![screen shot 2015-07-19 at 7 13 58 pm](https://cloud.githubusercontent.com/assets/1270099/8768549/5db36700-2e4a-11e5-96f8-0532452e2c42.png)

Posts that link to the blob. It needs a composer up top, so you can make a post about it.

![screen shot 2015-07-19 at 7 14 05 pm](https://cloud.githubusercontent.com/assets/1270099/8768550/5db37e2a-2e4a-11e5-92fc-6a9897ef30d6.png)

Basically, view source. We can expand on this, to let you "Save As" somewhere local, edit, and publish. Also, needs to recognize a large binary file (eg images) and do something other than trying to dump the data into there. Ugh.

**Comment by pfraze** [comment]
Ok, check out this flow. I open a file on my computer, then publish it on my feed:

![screen shot 2015-07-20 at 2 10 25 pm](https://cloud.githubusercontent.com/assets/1270099/8784860/a3e04032-2ee9-11e5-868f-0dacf37c0ba9.png)
![screen shot 2015-07-20 at 2 11 29 pm](https://cloud.githubusercontent.com/assets/1270099/8784859/a3dedd64-2ee9-11e5-8c24-0b9a48c48e2c.png)
![screen shot 2015-07-20 at 2 11 35 pm](https://cloud.githubusercontent.com/assets/1270099/8784862/a3e0aad6-2ee9-11e5-9c18-9f87c27da4f4.png)
![screen shot 2015-07-20 at 2 11 41 pm](https://cloud.githubusercontent.com/assets/1270099/8784861/a3e06ca6-2ee9-11e5-94db-eccb73c9fb9a.png)
![screen shot 2015-07-20 at 2 12 05 pm](https://cloud.githubusercontent.com/assets/1270099/8784863/a3e2a23c-2ee9-11e5-8b2d-3b0388a74129.png)
![screen shot 2015-07-20 at 2 12 07 pm](https://cloud.githubusercontent.com/assets/1270099/8784864/a3e34c00-2ee9-11e5-8203-8ee5ab4be80d.png)
![screen shot 2015-07-20 at 2 12 11 pm](https://cloud.githubusercontent.com/assets/1270099/8784865/a3eca2fa-2ee9-11e5-9727-8f3b382e32be.png)
![screen shot 2015-07-20 at 2 12 15 pm](https://cloud.githubusercontent.com/assets/1270099/8784866/a3eef302-2ee9-11e5-88e5-ebdc8a1e6604.png)
![screen shot 2015-07-20 at 2 12 30 pm](https://cloud.githubusercontent.com/assets/1270099/8784867/a3f45158-2ee9-11e5-9ce0-c938509cb494.png)

**Comment by dominictarr** [comment]
looks good!

**Comment by pfraze** [comment]
Next up on my todo list is a search page. It'll be pretty simple, no prebuilt indexes. It'll filter posts and users, matching against (respectively) post text and filenames, and usernames.

The editor panel only works for plaintext files at the moment... if it's binary, you get the garbled plaintext rendering. What would be very neat is, to detect the filetype and give some simple editors.

 - Images: filters, cropping, resizing, add text
 - Videos, Audio: recut, import clips, convert format (eg to gif)
 - Other: dunno, maybe a hex editor?

On the topic of HTML publishing, a few thoughts:

 - On publish, we can run the extraction process to detect all external resources, add them as attachments, and replace their references in the HTML. (Or, if we do a "bundle namespace," we can leave the references in the HTML, so long as they're relative.)
 - After publish, users can attach lots of data and metadata (by linking to the blob). Pages can exploit this by fetching links to self, if they're aware of their own hash. Patchwork can exploit this for, well, anything.
 - There are a *lot* of novel APIs we can make available to these pages: SSB, user profiles, encryption, p2p connections, HD access, payments... we should start fantasizing about this!

**Comment by pfraze** [comment]
Also missing right now, is fine-grained control over attachment metadata, and also links between forks. A forked file should link back to its original for sure! Then we can have a sidepanel showing all the different versions in a tree, and let you diff them

**Comment by pfraze** [comment]
![screen shot 2015-07-20 at 5 29 33 pm](https://cloud.githubusercontent.com/assets/1270099/8789075/fe07f77e-2f04-11e5-90db-7c606e4863f1.png)

**Comment by dominictarr** [comment]
is the data model the most important thing here? or the dev experience?
we can use code from @substack's htmlbin (well it's probably in a module) to bundle the file into one blob.
Speaking for my self, and many devs i think, i'd rather edit my app in $EDITOR than in a web page.
and of course, you'd need to do testing before publishing.
I'm imagining pointing ssb at a file and it watches it for changes or else press ctrl-shift-r as on a web browser. and when you are happy, hit publish.

**Comment by pfraze** [comment]
Yeah I agree about the editor situation -- this supports both options. It's exactly the flow you describe.

Im going to stick Ace or CodeMirror in the editor pane, because it doesn't cost us much to do and even advanced users would appreciate a good inline editor. Rookies would especially appreciate it!

**Comment by dominictarr** [comment]
yes, agreed.

**Comment by pfraze** [comment]
Started with codemirror but it wasnt a magic dropin, just used Behave.js to get nice tabbing behaviors.

**Comment by pfraze** [comment]
I bet something like http://fluidapp.com/ would be pretty easy to pull off

**Comment by dominictarr** [comment]
yup, it would be great to be able to have ui-less node apps too.
